src/ann_solo/utils.py

```diff
@@ -1,104 +1,338 @@
-            ssms, key=operator.attrgetter('search_engine_score'), reverse=True,
-            is_decoy=operator.attrgetter('is_decoy'), remove_decoy=True,
-            formula=1, correction=0, full_output=True):
-        ssm.q = q
-        if q <= fdr:
-            yield ssm
+    logging.debug(
+        "Compute features for semi-supervised scoring from %d SSMs", len(ssms)
+    )
+    # Create a Dataset with SSM features.
+    features = pd.DataFrame(_compute_ssm_features(ssms))
+    features["group"] = _get_ssm_groups(ssms, min_group_size) if grouped else 0
+    dataset = mokapot.dataset.LinearPsmDataset(
+        features,
+        target_column="is_target",
+        spectrum_columns="index",
+        peptide_column="sequence",
+        group_column="group",
+    )
+    if grouped:
+        logging.debug(
+            "Partitioned %d SSMs into %d groups",
+            len(ssms),
+            features["group"].nunique(),
+        )
+    # Define the mokapot model.
+    #   - Choice between a random forest, linear SVM, or no semi-supervised
+    #     learning.
+    #   - Features are preprocessed by standardizing them, removing
+    #     zero-variance features, and removing highly correlated features.
+    #   - We perform minimal tuning of TODO hyperparameters.
+    if model is None:
+        logging.debug("Calculate q-values based on the cosine similarity")
+        # Calculate q-values based on the cosine similarity.
+        confidences = dataset.assign_confidence(features["cosine"], True)
+    else:
+        logging.debug(
+            "Train semi-supervised %s model and score SSMs", model.upper()
+        )
+        scaler = make_pipeline(
+            StandardScaler(),
+            VarianceThreshold(),
+            CorrelationThreshold(0.95),
+        )
+        if model == "svm":
+            clf = mokapot.model.PercolatorModel(scaler, train_fdr=fdr)
+        elif model == "rf":
+            clf = mokapot.Model(
+                GridSearchCV(
+                    RandomForestClassifier(random_state=1),
+                    param_grid={
+                        "max_depth": [3, 5, 7, 9, None],
+                        "class_weight": [
+                            {0: neg, 1: pos}
+                            for neg in (0.1, 1, 10)
+                            for pos in (0.1, 1, 10)
+                        ],
+                    },
+                    refit=False,
+                    cv=3,
+                    n_jobs=-1,
+                ),
+                scaler,
+                train_fdr=fdr,
+            )
         else:
-            break
+            raise ValueError(
+                "Unknown semi-supervised machine learning model given"
+            )
+        # Train the mokapot model and combine the SSMs for all groups.
+        confidences, _ = mokapot.brew(dataset, clf, fdr, folds=5)
+    ssm_scores = pd.concat(
+        [
+            confidences.group_confidence_estimates[group].psms[
```

Comment:
I would suggest making a small change how ANN-SoLo reports results to include all SSMs, not only those below the specified FDR threshold. That can be useful for downstream analyses. In that case, maybe doing the filtering at a higher level in the code will give a better design.

Reply:
We can discuss it tomorrow whether we want to include all True Targets at the end or just the matches passing the FDR threshold. It will also be hard to benchmark now that we have all results of other tools. I believe they always return the best matching true targets above the threshold at the end. But we can discuss a better solution.